providers/base/bin/snap_confinement_test.py

```diff
@@ -17,64 +17,173 @@
 # You should have received a copy of the GNU General Public License
 # along with Checkbox.  If not, see <http://www.gnu.org/licenses/>.
 
+import argparse
 import json
 import logging
+import re
 import sys
 from checkbox_support.snap_utils.snapd import Snapd
 
-features_should_include = [
-    "kernel:caps",
-    "kernel:dbus",
-    "kernel:domain",
-    "kernel:file",
-    "kernel:mount",
-    "kernel:namespaces",
-    "kernel:network",
-    "kernel:ptrace",
-    "kernel:signal",
-    "parser:unsafe",
-]
 
+class SystemConfinement:
```

Comment:
Both classes in this testing program have only one method and they don't benefit from being a class. Simple function would be cleaner (for instance the indentation would be better allowing cleaner code).

Furthermore the function could be decomposed further to facilitate unit testing. For instance the "check_snap_confinment" could accept snap's data as an argument, and do the checks. Then in the main scope in the  iterating over snaps could be done and this function could be invoked for each.
In turn the unit test becomes simple - the snap data would have to be prepared with the conditions you want, and so on...

Reply:
Honestly, I was referring to the structure [snappy tests](https://github.com/canonical/checkbox/blob/main/providers/base/bin/snap_tests.py).

I think the idea is to separate the test cases by classes, and the data like `features_should_include` or `allowlist_snaps` could be located in that class. Re-write it to only functions are OK, but I think both ways have their own advantages.

Regarding decomposing the functions to apply unit testing, there are 2 things I'd like to discuss with you.
1. For this test case, I feel the unit tests could be implemented for functions in [snapd.py](https://github.com/canonical/checkbox/blob/main/checkbox-support/checkbox_support/snap_utils/snapd.py). As this test is calling the functions in snapd.py and get data in the dictionaries, it seems to me that unit tests for this test case are not necessary.
2. In my opinion, wrapping those `if conditions` to functions or not are fine. If the functions will not be reused or extended, keeping these simple `if conditions` will be easier to understand the test steps. If we are going to wrap them in functions like `check_snap_confinement()` or `check_snap_demode()`, it will be better to keep the class structure and implement these functions in classes. Something like,
```python3
class SnapsConfinement:
    def __init__(self):
        data = Snapd.list()

    def check_snap_confinement(self, snap, snap_name):
        confinement = snap.get("confinement")
        if confinement != "strict":
            logging.error(
                "Snap '%s' confinement is expected to be 'strict' "
                "but got '%s'", snap_name, confinement,
            )
            return False
        return True
```
And iterate the snaps `self.data` and get the snap name in `invoked()` functions.

Which way do you prefer?

Reply:
>keeping these simple if conditions will be easier to understand the test steps.

I agree with this statement.

If what you want to have is a class that reflects one idea of a test, like check "check snaps' confinement", then I think natural decomposition would be to iterate over the snaps in `invoked` and call the `is_snap_confined` check for each (btw. I changed the name on purpose - for checks that end in booleans the is_something is clearer tbh).
I hope this answers your question.

Reply:
Thank you for the clarification.

I rethought it a little bit. Since these 2 test classes are simple, I would prefer just to flatten them into functions `test_system_confinement()` and `test_snaps_confinement()`. And the attributes `features_should_include` and `allowlist_snaps` for each class become a local variable in each function.

To keep the test steps easy to understand, I would leave the `if conditions` as they are.
Feel free to give any comments in case I misunderstand anything.